src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -289,6 +291,191 @@ CitusGenerateDeferredQueryStrings(CustomScanState *node, EState *estate, int efl
 	{
 		HandleDeferredShardPruningForFastPathQueries(distributedPlan);
 	}
+
+	if (jobQuery->commandType != CMD_SELECT)
+	{
+		/* prevent concurrent placement changes */
+		AcquireMetadataLocks(workerJob->taskList);
+
+		/* modify tasks are always assigned using first-replica policy */
+		workerJob->taskList = FirstReplicaAssignTaskList(workerJob->taskList);
+	}
+
+	if (list_length(distributedPlan->workerJob->taskList) != 1)
+	{
+		/*
+		 * We might have zero shard queries or multi-row INSERTs at this point,
+		 * we only want to cache single task queries.
+		 */
+		return;
+	}
+
+	/*
+	 * As long as the task accesses local node and the query doesn't have
+	 * any volatile functions, we cache the local Postgres plan on the
+	 * shard for re-use.
+	 */
+	Task *task = linitial(distributedPlan->workerJob->taskList);
+	if (EnableLocalExecution && TaskAccessesLocalNode(task) &&
+		!contain_volatile_functions(
+			(Node *) originalDistributedPlan->workerJob->jobQuery))
+	{
+		CacheLocalPlanForTask(task, originalDistributedPlan);
+	}
+	else
```

Comment:
@marcocitus does the else part make sense to you as well?

Reply:
yes, the comments are helpful